hudi-common/src/main/java/org/apache/hudi/common/util/ClusteringUtils.java

```diff
@@ -365,20 +370,56 @@ public static Option<HoodieInstant> getEarliestInstantToRetainForClustering(
           // TODO: This case has to be handled. HUDI-6352
           retainLowerBound = cleanInstant.getTimestamp();
         }
-
-        oldestInstantToRetain = replaceOrClusterTimeline.filter(instant ->
-                HoodieTimeline.compareTimestamps(
-                    instant.getTimestamp(),
-                    HoodieTimeline.GREATER_THAN_OR_EQUALS,
-                    retainLowerBound))
-            .firstInstant();
+        oldestInstantToRetain = replaceOrClusterTimeline.findInstantsAfterOrEquals(retainLowerBound).firstInstant();
       } else {
         oldestInstantToRetain = replaceOrClusterTimeline.firstInstant();
       }
     }
+    if (!shouldArchiveBeyondSavepoint) {
+      // explicitly check the savepoint timeline and guard against the first one
+      Option<HoodieInstant> firstSavepointOpt = activeTimeline.getSavePointTimeline().filterCompletedInstants().firstInstant();
+      oldestInstantToRetain = Option.ofNullable(HoodieTimeline.minTimestampInstant(oldestInstantToRetain.orElse(null), firstSavepointOpt.orElse(null)));
+    }
     return oldestInstantToRetain;
   }
 
+  public static Option<String> getEarliestReplacedSavepointInClean(HoodieActiveTimeline activeTimeline, HoodieCleaningPolicy cleanerPolicy,
```

Comment:
we can name this method as 
getEarliestCommitToNotArchive. and keep this method in HoodieTimelineArchiver. 

either we can compute and pass the value to ClusteringUtils.getEarliestInstantToRetainForClustering  or make a function as an arg.

Reply:
`getEarliestCommitToNotArchive` is not a good name because all the logic in archiver is to compute the "earliest commit to not archive".